### PR TITLE
Improvements for PaginatorHelper

### LINKED ISSUE #9503: Could not generate pagination links with passed parameters in route.
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.3
* Platform and Target: Apache 2.4, PHP 7.0.10, MySQL

### What you did

Configured this route:

```
Router::scope('/:lang', [
    'lang' => '[a-z][a-z]'
    ], function (RouteBuilder $routes) {
    $routes->connect('/explore', [
        'controller' => 'Explore',
        'action' => 'index'
    ], [
        '_name' => 'explore',
        'pass' => ['lang']
    ]);
});
```

Tried to generate pagination numbers:

```
echo $this->Paginator->numbers(['url' => ['lang' => 'en']]);
```

Triggered this missing route error:

```
Error: A route matching "array ( 'page' => false, 0 => 'en', 'lang' => 'en', 'action' => 'index', 'plugin' => NULL, 'controller' => 'Explore', '_ext' => NULL, )" could not be found.
```

Notice that both the passed parameter `lang` is defined twice. Once as `0=>'en'` and `'lang'=>'en'`.

### Expected Behavior

It should not be inject named parameters as a numeric key.

### Actual Behavior

The source of the problem was in the PaginatorHelper constructor.

```
        $this->config(
            'options.url',
            array_merge($this->request->params['pass'], ['?' => $query])
        );
```
The above is attempting to persist the current route parameters. Which is a good feature, but I think injecting `params['pass']` is the wrong approach. Since those are numeric key/pair set.

This raises the question of where the bug is located. Should the `params['pass']` contain keys with the name of the parameter instead of numeric keys, or should the helper's constructor inject named parameter manually (some how).

The work around for the above problem was to erase the options URL before generating links.

```
        $this->Paginator->config('options.url',[], false);
```

### PR DESCRIPTION
**Pull request by thinkingmedia**
Adds the following to the helper
- moved core of generateUrl() to buildUrl() and returns a raw URL array of pagination parameters.
- generateUrl calls buildUrl()
- added total() which returns page count.
- added to meta() support for generating first and last meta links

The object of the PR is to allow developers to generate pagination URLs in their array format for easier modification outside of the helper.

closes #9503

### CODE REVIEW

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -1110,6 +1155,8 @@ public function meta(array $options = [])
             $options['block'] = __FUNCTION__;
         }
         $this->_View->append($options['block'], $out);
+
+        return null;
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: If you invert the logic, you will not have to do "return null;" at the end of a method.

File: src/View/Helper/PaginatorHelper.php [hunk 2]
```diff
@@ -493,6 +507,21 @@ public function generateUrl(array $options = [], $model = null, $urlOptions = fa
...
+    }
+
+    /**
+     * Merges passed URL options with current pagination state to generate a pagination URL.
+     *
+     * @param array $options Pagination/URL options array
+     * @param string|null $model Which model to paginate on
+     * @return array An array of URL parameters
+     */
+    public function buildUrl(array $options = [], $model = null)
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: Do we need to expose this as public method?

**Comment by markstory** [comment]
Should this be targetting `3.next` instead?

File: src/View/Helper/PaginatorHelper.php [hunk 3]
```diff
@@ -195,6 +195,23 @@ public function current($model = null)
     }
 
     /**
+     * Gets the total number of pages in the recardset for the given model.
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: typo in record.

File: src/View/Helper/PaginatorHelper.php [hunk 4]
```diff
@@ -195,6 +195,23 @@ public function current($model = null)
...
+     */
+    public function total($model = null)
+    {
+        $params = $this->params($model);
+
+        if (isset($params['pageCount'])) {
+            return $params['pageCount'];
+        }
+
+        return 1;
```
Comment (markstory) [inline src/View/Helper/PaginatorHelper.php]: Need tests for this method.

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 5]
```diff
@@ -2451,110 +2458,68 @@ public function testWithZeroPages()
...
+            // Verifies that the last page only has a prev link.
+            [2, true, false, 2, [], '<link href="http://localhost/index" rel="prev"/>'],
+            // Verifies that the last page only has a prev, first and last link.
+            [2, true, false, 2, ['first' => true, 'last' => true], '<link href="http://localhost/index" rel="prev"/><link href="http://localhost/index" rel="first"/><link href="http://localhost/index?page=2" rel="last"/>'],
+            // Verifies that a page in the middle has both links.
+            [5, true, true, 10, [], '<link href="http://localhost/index?page=4" rel="prev"/><link href="http://localhost/index?page=6" rel="next"/>'],
+                // Verifies that a page in the middle has both links.
+            [5, true, true, 10, ['first' => true, 'last' => true], '<link href="http://localhost/index?page=4" rel="prev"/><link href="http://localhost/index?page=6" rel="next"/><link href="http://localhost/index" rel="first"/><link href="http://localhost/index?page=10" rel="last"/>']
```
Comment (markstory) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: Any chance we could get these really long lines wrapped somehow?

File: src/View/Helper/PaginatorHelper.php (hunk 2 above)
Comment (thinkingmedia) [inline src/View/Helper/PaginatorHelper.php]: it doesn't technically return a URL. it returns an array, so how about calling it `addUrlParams()`? Which is closer to what it does. (or something like that).

**Comment by thinkingmedia** [comment]
@markstory sure, 3.next is fine. Do I have to rebase it?

**Comment by dereuromark** [comment]
No, just switch the branch on the top (dropdown in edit mode).

File: tests/TestCase/View/Helper/PaginatorHelperTest.php (hunk 5 above)
Comment (thinkingmedia) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: @markstory Yes, these lines can be wrapped by using a carriage return.

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 6]
```diff
@@ -2394,6 +2394,20 @@ public function testCurrent()
...
+     *
+     * @return void
+     */
+    public function testTotal()
+    {
+        $result = $this->Paginator->total();
+        $this->assertEquals($this->Paginator->request->params['paging']['Article']['pageCount'], $result);
+
+        $result = $this->Paginator->total('Incorrect');
+        $this->assertEquals(0, $result);
```
Comment (dereuromark) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: Let's please use assertSame() to be type safe.

Comment (thinkingmedia) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: didn't know that. thanks.

**Comment by markstory** [comment]
You might need to rebase @thinkingmedia

**Comment by thinkingmedia** [comment]
This got all screwed up. I think I'll just open a new PR.

**Pull request closed without merging**